docs/proposals/config_data.md

```diff
@@ -0,0 +1,298 @@
+once it is made visible (or perform a rolling-update on consumers of that object. This is especially true if ``ConfigData`` is injected as environment variables. Dynamic configuration with environment variables is out of scope).
+
+### Advantages
+
+* Easy distribution through the API server.
+* Provides configuration data in a consumer-agnostic manner.
+* Persistent configuration information through API versioning.
+* Ability to use ``Kubernetes`` authentication and security policies with configuration objects.
+* By leveraging the power of ``/watch`` API or watching the volume, container processes can implement responsiveness to configuration changes.
+
+## API Resource
+
+A new resource for ``ConfigData`` will be added to the ``Extentions`` API Group:
+
+```go
+	type ConfigData struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		Data map[string]string `json:"data,omitempty"`
+	}
+
+	type ConfigDataList struct {
+		TypeMeta `json:",inline"`
+		ListMeta `json:"metadata,omitempty"`
+
+		Items []ConfigData `json:"items"`
+	}
+```
+
+A new API resource selector (``ConfigDataSelector``) will be added to select objects and fields from those objects
+for consuming API resources through the downward API.
+
+``Registry`` interface will be added to accompany the ``ConfigData`` resource.  
+
+### Volume Source
+
+A new ``ConfigDataVolumeSource`` type of volume source containing the ``ConfigData`` object will be added to the ``VolumeSource`` struct in the API:
+
+```go
+	type ConfigDataVolumeSource struct {
+		ConfigDataName string `json:"configDataName"`
+	}
+```
+
+**Note:** The update logic used in the downward API volume plug-in will be extracted and re-used in the volume plug-in for ``ConfigData``.
+
+## Creating ConfigData
+The following file ``pod-config-example.json`` contains data intended to be consumed as environment variables.
+```json
+{
+    "apiVersion": "extensions/v1beta1",
+    "metadata": {
+        "name": "pod-config-example",
+    },
+    "kind": "ConfigData",
+    "data": {
+        "server_address": "10.240.13.14:8998",
+        "db_address": "10.240.55.3",
+        "db_port": "3306"
+    }
+}
+```
+
+To persist this ConfigData object to the API server, use the following command:
+```bash
+kubectl create -f pod-config.json
+```
+
+Now, the ``pod-config-example`` ConfigData object is stored in ``etcd``.
+The new object can be accessed, referenced, used and updated through the API server.
+
+## Configuration Workflow
+
+### Environment Variables
+
+The component configuration workflow is given below:
+
+1. ``ConfigData`` object is created through the API server.
+2. New components using the ``ConfigData`` object are created.
+3. ``ConfigData`` object is retrieved using the downward API.
+4. ``Data`` key/value pairs of ``ConfigData`` will be exposed as environment variables (``EnvVarSource``).
+5. API server is watched for modifications to the object.
```

Comment:
This actually isn't true.  We can't change the values of env vars without restarting containers.  So, if you want an eventually consistent view, you need to use the volume plugin.